### PR TITLE
Add more unit tests for IsoRoute class

### PR DESCRIPTION
**Pull request by quinton-hoole**
### Summary

Add unit tests for 13 more methods of IsoRoute class.

### Note

All commits except the last one can be ignored as far as code review goes.  The other commits are covered by separate PR's, and will be merged before this one, so will disappear from this PR.

### Coverage

Test coverage is somewhat improved, most notably:

```
make coverage
...
File 'IsoRoute.cpp'
Lines executed:29.05% of 771

File 'LineBufferOverlay.cpp'
Lines executed:63.95% of 172

File 'Polar.cpp'
Lines executed:69.87% of 634

File 'PolygonRegion.cpp'
Lines executed:61.07% of 298

File 'Position.cpp'
Lines executed:37.54% of 301

File 'RouteMap.cpp'
Lines executed:3.09% of 485

File 'RoutePoint.cpp'
Lines executed:10.40% of 327

File 'RoutingTablePanel.cpp'
Lines executed:12.63% of 1378

File 'SettingsDialog.cpp'
Lines executed:18.62% of 145

File 'Utilities.cpp'
Lines executed:98.15% of 54

File 'WeatherDataProvider.cpp'
Lines executed:20.44% of 225

```

cc @sebastien-rosset @rgleason

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -1,10 +1,13 @@
-cmake_minimum_required(VERSION 3.14)
+cmake_minimum_required(VERSION 3.15.0)
 
 project(weather_routing_pi_tests)
 set(CMAKE_CXX_STANDARD 17)
 message(STATUS "Building weather routing plugin tests")
+# make sure to use a Debug build, for coverage
+set(CMAKE_BUILD_TYPE "Debug")
```
Comment (sebastien-rosset) [inline test/CMakeLists.txt:7]: We should not force doing a Debug build.

Comment (quinton-hoole) [reply test/CMakeLists.txt:7]: This setting is inside [test/CMakeLists.txt](https://github.com/rgleason/weather_routing_pi/pull/448/files/72a0dcbc7a0dd0ff729830f274e65e2b0be23066#diff-33394812ba204689144fd2f80832db83853ba1cb32403edb4e15fe4893e675fd) and therefore only affects the build type for tests.  Whenever tests are built, the test binary should always have debug symbols included.

**Comment by quinton-hoole** [comment]
> I think first we need to complete the review of #445 and merge, then do a full review of this PR.

Thanks @sebastien-rosset . #445 is ready for merge now, IMO. Then you can proceed with the review of this one.

**Comment by sebastien-rosset** [comment]
I merged #445, and now there are merge conflicts in this PR. Can you resolve them?

**Comment by quinton-hoole** [comment]
> I merged #445, and now there are merge conflicts in this PR. Can you resolve them?

Yes, sure, will do now.

File: test/IsoRoute_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+       m_position[0]->next = m_position[1];
+       for(int i = 1; i < sizeof(m_position)/sizeof(Position*) -1; ++i) {
+         m_position[i]->prev = m_position[i-1];
+         m_position[i]->next = m_position[i+1];
+       }
+       m_isoRoute = new IsoRoute(m_position[0]->BuildSkipList(), 1);
+    }
+ 
+   ~IsoRouteTest() override {
+     // You can do clean-up work that doesn't throw exceptions here.
```
[the remark this answers is not in the document]
Comment (quinton-hoole) [reply test/IsoRoute_tests.cpp]: Already dealt with [here](https://github.com/rgleason/weather_routing_pi/pull/445#discussion_r2173544422)

File: test/IsoRoute_tests.cpp [hunk 3]
```diff
@@ -51,7 +51,125 @@
...
+  m_isoRoute->Print();
+ }
+
+ TEST_F(IsoRouteTest, PrintSkipBasic) {
+  // Check that the PrintSkip method does not crash.
+  m_isoRoute->PrintSkip();
+ }
+
+ TEST_F(IsoRouteTest, ClosestPositionBasic) {
+  // Check that the closest position is the first one in the list.
```
Comment (sebastien-rosset) [inline test/IsoRoute_tests.cpp:75]:
```suggestion
  // When querying with exact coordinates of an existing position,
  // should return that exact position
```

File: test/IsoRoute_tests.cpp [hunk 4]
```diff
@@ -51,7 +51,125 @@
...
+  Position* closest = m_isoRoute->ClosestPosition(m_position[0]->lat, m_position[0]->lon);
+  EXPECT_FLOAT_EQ(closest->lat, m_position[0]->lat);
+  EXPECT_FLOAT_EQ(closest->lon, m_position[0]->lon);
+  // Check that the closest position to a point not in the list is the last one (in this case).
```
[the remark this answers is not in the document]
Comment (quinton-hoole) [reply test/IsoRoute_tests.cpp:79]: Yes, that's why I wrote "(in this case)".  But I can reword the comment.

File: test/IsoRoute_tests.cpp (hunk 3 above)
Comment (quinton-hoole) [reply test/IsoRoute_tests.cpp:75]: Sure, happy to change.

File: test/IsoRoute_tests.cpp [hunk 5]
```diff
@@ -51,7 +51,126 @@
...
+
+ TEST_F(IsoRouteTest, ContainsRouteBasic) {
+  // Check that the route does not contains itself.
+  EXPECT_FALSE(m_isoRoute->ContainsRoute(m_isoRoute));
+ }
+
+ TEST_F(IsoRouteTest, FindIsoRouteBoundsBasic) {
+  double bounds[4];
+  // Check that the bounds are found correctly.
+  m_isoRoute->FindIsoRouteBounds(bounds);
```
Comment (sebastien-rosset) [inline test/IsoRoute_tests.cpp:100]: These tests are quite fragile/brittle and would easily break when the test setup is changed, such as when the route becomes a quadrilateral (or some other shape) instead of triangle.

Also, for each test you have to keep going back and forth between the test and the setup code to understand why the test succeeds or fails.

It hard to maintain: Future developers can't easily understand what should happen when the test fails.
It also yields false confidence: Passes for wrong reasons (coincidental data alignment)

File: test/IsoRoute_tests.cpp [hunk 6]
```diff
@@ -51,7 +51,126 @@
...
+  EXPECT_FLOAT_EQ(closest->lon, m_position[2]->lon);
+ }
+
+ TEST_F(IsoRouteTest, CompletelyContainedBasic) {
+  // Check that the route does not completely contains itself.
+  EXPECT_FALSE(m_isoRoute->CompletelyContained(m_isoRoute));
+ }
+
+ TEST_F(IsoRouteTest, ContainsRouteBasic) {
+  // Check that the route does not contains itself.
```
[the remark this answers is not in the document]
Comment (sebastien-rosset) [reply test/IsoRoute_tests.cpp:93]: Hmm, that means we are writing tests for the sole purpose of asserting the current implementation, without understanding why the code works that way.
As I have mentioned in my other comment, it's a bit hard to follow what the tests do, because everywhere it states "basic". Instead the name should describe what it does.

My minimum bar for writing unit tests is that the tests should be written using the TDD principle. I.e. the person writing the test should write the assertions based on their expectation of what the code MUST do, not based on the implementation. Otherwise, we commit a bunch of tests, but who is actually going to do the work to review all of these tests in details? And then later, given a library of tests, how will you know whether the tests were written using the TDD approach, or just matching the existing implementation?

File: test/IsoRoute_tests.cpp [hunk 7]
```diff
@@ -51,7 +51,126 @@
...
+  closest = m_isoRoute->ClosestPosition(outsidePos.lat, outsidePos.lon);
+  EXPECT_FLOAT_EQ(closest->lat, m_position[2]->lat);
+  EXPECT_FLOAT_EQ(closest->lon, m_position[2]->lon);
+ }
+
+ TEST_F(IsoRouteTest, CompletelyContainedBasic) {
+  // Check that the route does not completely contains itself.
+  EXPECT_FALSE(m_isoRoute->CompletelyContained(m_isoRoute));
+ }
+
```
[the remark this answers is not in the document]
Comment (sebastien-rosset) [reply test/IsoRoute_tests.cpp:91]: IMO, test names should have a structure like this: `[MethodName][SpecificCase]`
There are so many ways these functions can have "basic" testing.

- Contains self
- Contains PartiallyOverlappingRoute
- Contains route with single point
- Square contains smaller square
- Square does not contain bigger shape

By using the word "basic" and later adding more tests like the one above, what will be the rule for deciding what is "basic"? For example, isn't "Square contains smaller square" test a pretty basic test?
And when you look at a long list of tests, some of which may fail, it's a LOT easier to troubleshoot if the failed test is named "RouteDoesNotContainSelf" rather than "ContainsRouteBasic".

[the remark this answers is not in the document]
Comment (quinton-hoole) [reply test/IsoRoute_tests.cpp:91]: > IMO, test names should have a structure like this: `[MethodName][SpecificCase]` There are so many ways these functions can have "basic" testing.
> 
> * Contains self
> * Contains PartiallyOverlappingRoute
> * Contains route with single point
> * Square contains smaller square
> * Square does not contain bigger shape
> 
> By using the word "basic" and later adding more tests like the one above, what will be the rule for deciding what is "basic"? For example, isn't "Square contains smaller square" test a pretty basic test? And when you look at a long list of tests, some of which may fail, it's a LOT easier to troubleshoot if the failed test is named "RouteDoesNotContainSelf" rather than "ContainsRouteBasic".

I don't necessarily agree, but I can change the names of the tests if you feel strongly about it.

**Comment by sebastien-rosset** [comment]
For the record, I disagree with the PR being merged as I was going to suggest changes. Some references:

1. Roy Osherove - "The Art of Unit Testing" 
2. Martin Fowler - Test Naming Patterns
3. Kent Beck - "Test Driven Development: By Example"

**Comment by quinton-hoole** [comment]
> For the record, I disagree with the PR being merged as I was going to suggest changes. Some references:
> 
> 1. Roy Osherove - "The Art of Unit Testing"
> 2. Martin Fowler - Test Naming Patterns
> 3. Kent Beck - "Test Driven Development: By Example"

I merged the PR by mistake, As I thought you had finished commenting. May I suggest that we address whatever changes you think to be needed in followup PR's?  I have attempted to respond to all of your comments above. See what you think.

**Comment by sebastien-rosset** [comment]
More feedback:
1. The weather routing plugin implements complex geometric and numerical algorithms. Unlike typical business logic or UI code, this plugin deals with computational geometry (Point-in-polygon tests, ray casting algorithms, polygon intersections...); Geographic calculations (Great circle distances, bearings, coordinate transformations); Optimization algorithms.
2. The core challenge is that geometric correctness cannot be validated by simple input/output matching - it requires understanding the mathematical properties and edge cases of spatial algorithms.
3. Lines of Code coverage creates a dangerous illusion of quality in this domain because the geometric algorithms have high cyclomatic complexity but shallow semantic coverage. For example, testing that `ContainsRoute(self)` returns `false` achieves line coverage but validates nothing about the ray-casting algorithm's correctness for:
   - Points inside/outside polygons
   - Points exactly on polygon edges
   - Vertices of the polygon
   - Numerical precision edge cases
4. Adding 13 shallow "does not crash/trivial" tests inflates coverage percentages while providing minimal assurance that the routing algorithms work correctly under real navigation scenarios.

Weather routing tests should validate correctness of complex algorithms, not just code execution paths. While I'm sure it's possible to achieve high LoC coverage with trivial tests, it will be an illusion. I'd rather start with lower LoC coverage but with tests that actually verify algorithmic correctness.

The fundamental problem with LoC coverage as a primary metric is that it answers the wrong question. It tells us "did we execute this code?" but not "does this code work correctly?" For geometric and optimization algorithms, these are vastly different questions. Supposed you've achieved 90% LoC coverage with a couple hundred unit tests. What do you plan after that? How will we know which tests are just testing LoC execution versus algorithmic correctness? How will contributors know what tests need to be improved and how? And when you make a change to code, what confidence will you have in the test coverage (LoC superficial or actual algorithmic correctness)?

File: test/IsoRoute_tests.cpp [hunk 8]
```diff
@@ -51,7 +51,126 @@
...
+  EXPECT_EQ(m_isoRoute->SkipCount(), sizeof(m_position)/sizeof(Position*) - 1);
+ }
+
+ TEST_F(IsoRouteTest, CountBasic) {
+  // Check that the count of positions in the route is correct.
+  EXPECT_EQ(m_isoRoute->Count(), sizeof(m_position)/sizeof(Position*));
+ }
+
+  // Minimal test to check that test compiles and Propagate does not crash.
+ TEST_F(IsoRouteTest, PropagateBasic) {
```
Comment (sebastien-rosset) [inline test/IsoRoute_tests.cpp:153]: The comment claims this tests that "Propagate does not crash," but it only tests one specific scenario: an empty/default RouteMapConfiguration with an empty IsoRouteList. This is like testing that a car engine starts with the ignition key, then claiming you've verified "the engine never fails."

## Problem 1: False Confidence Through Narrow Input Space

The Propagate function has dozens of failure modes:

- Invalid weather data
- Malformed boat polar diagrams
- Degenerate route geometries
- Out-of-bounds coordinates
- Conflicting constraint configurations
- Memory allocation failures
- Numerical overflow/underflow

Testing with default-constructed objects exercises none of these realistic failure scenarios.

## Problem 2: "Does Not Crash" Is Not a Test Goal

Not crashing is a basic expectation of every line of code, not a specific test objective. Real tests verify functional correctness, and non-crashing is an implicit prerequisite.

## Problem 3: Comment Misrepresents Test Scope

The comment creates a false sense of algorithmic coverage. Someone reading this might think "great, Propagate is tested" and move on, not realizing that the core weather routing algorithm has zero meaningful validation.

A more honest comment would be:
```c++
// TODO: This is a stub test that only verifies compilation.
// MISSING: Tests for weather routing correctness, edge cases, and realistic scenarios.
TEST_F(IsoRouteTest, PropagateCompilationStub) {
```

## What This Test Should Actually Be

If you're going to test Propagate, test its routing algorithm correctness.

## The Broader Problem

We should not create the appearance of testing without the substance. It's dangerous in algorithmic code because:

- It inflates coverage metrics without improving code quality
- It creates false confidence that critical algorithms are validated
- It establishes bad precedent for other contributors to write similarly hollow tests

**Comment by quinton-hoole** [comment]
Thank you for your feedback.
